File: Java/Tema5_Ejercicio6.java
/*
 * TEMA 5: ARREGLOS Y LISTAS - EJERCICIO 6
 * ========================================
 * Escriba un programa que busque un elemento en un arreglo.
 */

import java.util.Scanner;
import java.util.Arrays;

public class Tema5_Ejercicio6 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        int[] numeros = {15, 32, 48, 21, 67, 89, 12, 45, 76, 34};
        
        System.out.println("=== BÚSQUEDA EN ARREGLO ===");
        System.out.println("Arreglo: " + Arrays.toString(numeros));
        
        System.out.print("\nBusca un número: ");
        int numBuscado = scanner.nextInt();
        
        // Búsqueda lineal
        int posicion = -1;
        for (int i = 0; i < numeros.length; i++) {
            if (numeros[i] == numBuscado) {
                posicion = i;
                break;
            }
        }
        
        if (posicion != -1) {
            System.out.println("Número encontrado en la posición: " + posicion);
        } else {
            System.out.println("Número no encontrado.");
        }
        
        // Búsqueda con Arrays.binarySearch() (requiere arreglo ordenado)
        int[] numerosOrdenados = numeros.clone();
        Arrays.sort(numerosOrdenados);
        int posicion2 = Arrays.binarySearch(numerosOrdenados, numBuscado);
        
        if (posicion2 >= 0) {
            System.out.println("Usando búsqueda binaria: posición " + posicion2 + 
                             " en el arreglo ordenado");
        }
        
        scanner.close();
    }
}

/*
 * SALIDA ESPERADA:
 * === BÚSQUEDA EN ARREGLO ===
 * Arreglo: [15, 32, 48, 21, 67, 89, 12, 45, 76, 34]
 * 
 * Busca un número: 67
 * Número encontrado en la posición: 4
 * Usando búsqueda binaria: posición 7 en el arreglo ordenado
 */

File: JS/Tema5_Ejercicio6.js
/*
 * TEMA 5: ARREGLOS Y LISTAS - EJERCICIO 6
 * ========================================
 * Búsqueda lineal y búsqueda binaria.
 * 
 * EXPLICACIÓN:
 * - Búsqueda lineal: recorre elemento por elemento
 * - Búsqueda binaria: divide y conquista (requiere array ordenado)
 * - indexOf() como alternativa nativa
 */

function busquedaLineal(arr, objetivo) {
    for (let i = 0; i < arr.length; i++) {
        if (arr[i] === objetivo) return i;
    }
    return -1;
}

function busquedaBinaria(arr, objetivo) {
    let izq = 0, der = arr.length - 1;
    
    while (izq <= der) {
        const mid = Math.floor((izq + der) / 2);
        if (arr[mid] === objetivo) return mid;
        if (arr[mid] < objetivo) izq = mid + 1;
        else der = mid - 1;
    }
    return -1;
}

const numeros = [10, 20, 30, 40, 50, 60, 70, 80, 90];
const objetivo = 40;

console.log("=== BÚSQUEDA EN ARRAYS ===");
console.log(`Array: ${numeros}`);
console.log(`Buscando: ${objetivo}`);

const indiceLineal = busquedaLineal(numeros, objetivo);
console.log(`\nBúsqueda lineal: índice ${indiceLineal}`);

const indiceBinaria = busquedaBinaria(numeros, objetivo);
console.log(`Búsqueda binaria: índice ${indiceBinaria}`);

console.log(`indexOf(): índice ${numeros.indexOf(objetivo)}`);

// No encontrado
console.log(`\nBuscando 55 (no existe):`);
console.log(`Búsqueda lineal: ${busquedaLineal(numeros, 55)}`);
console.log(`Búsqueda binaria: ${busquedaBinaria(numeros, 55)}`);

/*
 * SALIDA ESPERADA:
 * === BÚSQUEDA EN ARRAYS ===
 * Array: 10,20,30,40,50,60,70,80,90
 * Buscando: 40
 * 
 * Búsqueda lineal: índice 3
 * Búsqueda binaria: índice 3
 * indexOf(): índice 3
 * 
 * Buscando 55 (no existe):
 * Búsqueda lineal: -1
 * Búsqueda binaria: -1
 */
